front: select op with map when add train

theocrsb opened the pull request:
-bold pr on hover (for PRs on same track)
-select pr on map and choose track

depends: https://github.com/OpenRailAssociation/osrd/pull/9800 and https://github.com/OpenRailAssociation/osrd/pull/10157
closes https://github.com/osrd-project/osrd-confidential/issues/777

theocrsb committed:
front: select op with map when add train

-bold pr on hover (for PRs on same track)
-select pr on map and choose track

Signed-off-by: theocrsb <theo_crosbie@yahoo.fr>

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -144,12 +145,12 @@ const Map = ({
   const onFeatureClick = (e: MapLayerMouseEvent) => {
     if (preventPointSelection) return;
-    const result = getMapMouseEventNearestFeature(e, { layersId: ['chartis/tracks-geo/main'] });
+    const result = getMapMouseEventNearestFeature(e);
```
As you have removed the layers which are clickable, I can select all lines and points (signals, dectectors etc)

clarani replied:
Actually, I think you can keep the layersId you can interact with and just add the operationalPoints layers like this:
```suggestion
    const result = getMapMouseEventNearestFeature(e, {
      layersId: [
        'chartis/tracks-geo/main',
        'chartis/osrd_operational_point/geo',
        'chartis/osrd_operational_point_yardname/geo',
        'chartis/osrd_operational_point_name_short/geo',
        'chartis/osrd_operational_point_name/geo',
        'chartis/osrd_operational_point_kp/geo',
      ],
    });
```
If you do so, the user will be able to interact with an op even if it is covered by a signal, or you will be able to interact with the track even if it is covered by electrification or speed limit

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -103,6 +103,7 @@ const Map = ({
 
   const [mapIsLoaded, setMapIsLoaded] = useState(false);
 
+  const [operationnalPointId, setOperationnalPointId] = useState<string | undefined>(undefined);
```
- add hovered in the name
- operational has only 1 n 😉
```suggestion
  const [hoveredOperationalPointId, setHoveredOperationalPointId] = useState<string | undefined>(undefined);
```

theocrsb replied:
Done

Uriel-Sautron commented on front/src/modules/pathfinding/helpers/getStepLocation.ts:
```diff
@@ -1,19 +1,35 @@
 const getStepLocation = (step: PathItemLocation): PathItemLocation => {
+  const trackReference: OperationalPointReference['track_reference'] =
+    !('track' in step) && step.track_reference && 'track_name' in step.track_reference
+      ? { track_id: undefined, track_name: step.track_reference.track_name }
```
```suggestion
      ? {track_name: step.track_reference.track_name }
```
Here track is always undefined.

theocrsb replied:
Done

Uriel-Sautron commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/RenderPopup.tsx:
```diff
@@ -13,11 +15,18 @@ import { editoastToEditorEntity } from 'applications/editor/data/api';
+import {
+  osrdEditoastApi,
+  type OperationalPoint,
+  type OperationalPointReference,
```
OperationalPointReference is not used

theocrsb replied:
Done

Uriel-Sautron commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/RenderPopup.tsx:
```diff
@@ -13,11 +15,18 @@ import { editoastToEditorEntity } from 'applications/editor/data/api';
 import { setPointIti } from 'modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/setPointIti';
-import type { PathStep } from 'reducers/osrdconf/types';
+import { type PathStep } from 'reducers/osrdconf/types';
+import { getPointCoordinates } from 'utils/geometry';
+import { omit } from 'lodash';
```
eslint warning 'import order' and 'empty  line'

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,44 +32,101 @@ type AddPathStepPopupProps = {
+      if (isOperationalPoint) {
+        const trackId = featureInfoClick.feature.properties?.track_id;
+
+        const clickedTrack = await getTrackEntity({
```
Can you remove getTrackEntity by getInfraObjectEntity since you use it to fetch operational points and track sections ?

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,44 +32,101 @@ type AddPathStepPopupProps = {
+          body: [trackId],
+        }).unwrap();
+
+        const { parts } = result[0].railjson as OperationalPoint;
```
You can reuse this const l115 and 116 (we should avoid `any`s ;) )
```suggestion
        const operationalPoint = result[0].railjson as OperationalPoint;
```

theocrsb replied:
You're right, it's done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,44 +32,101 @@ type AddPathStepPopupProps = {
+          tracks: trackPartCoordinates,
+        });
+      } else {
+        setClickedOp(undefined);
```
I think it would be nice to reset clickedOp and offset before calling `calculateOffset` in the hook.

Can you rename the function by the way ?

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -1,7 +1,10 @@
+/* eslint-disable @typescript-eslint/no-unsafe-return */
 /* eslint-disable @typescript-eslint/no-explicit-any */
```
I think you can remove the `eslint-disable no-explicit-any` :)

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,37 +32,59 @@ type AddPathStepPopupProps = {
   const origin = useSelector(getOrigin);
   const destination = useSelector(getDestination);
 
+  const isOperationalPoint = useMemo(() => {
+    const properties = featureInfoClick?.feature?.properties;
+    return !!properties?.track_id || !!properties?.track_name;
+  }, [featureInfoClick]);
```
Suggestion: can we store in featureInfoClick if it is an operationalPoint or a track ?

This way, you won't have to compute `isOperationalPoint` and you will just have to do `featureInfoClick.isOperationalPoint` directly

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,37 +32,59 @@ type AddPathStepPopupProps = {
         return;
       }
 
+      setClickedOp(undefined);
```
You don't need to reset clickedOp here since you have already done it l138
```suggestion
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -27,37 +32,59 @@ type AddPathStepPopupProps = {
       }
 
+      setClickedOp(undefined);
+      // if operationnalPoint we already have coordinates
```
I think this comment is not useful anymore
```suggestion
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -67,49 +94,127 @@ function AddPathStepPopup({
-      trackNumber: trackProperties.extensions_sncf_track_number,
-    },
-  };
+  let pathStepProperties: PathStep;
+  if (isOperationalPoint && clickedOp && selectedTrack) {
+    const newPathStep: PathStep = {
+      ...omit(clickedOp, ['tracks']),
+      coordinates: selectedTrack.coordinates,
+      track_reference: selectedTrack.trackName
+        ? { track_name: selectedTrack.trackName }
+        : undefined,
+    };
+    pathStepProperties = {
+      ...newPathStep,
+    };
+  } else {
+    pathStepProperties = {
+      id: nextId(),
+      coordinates,
+      track: trackProperties.id,
+      offset: Math.round(trackOffset), // offset needs to be an integer
+      kp: trackProperties.kp,
+      metadata: {
+        lineCode: trackProperties.extensions_sncf_line_code,
+        lineName: trackProperties.extensions_sncf_line_name,
+        trackName: trackProperties.extensions_sncf_track_name,
+        trackNumber: trackProperties.extensions_sncf_track_number,
+      },
+    };
+  }
```
Suggestion: can you add a state variable like this:
```
const [newPathStep, setNewPathStep] =
    useState<Extract<PathStep, { track: string } | { uic: number }>>();
```

This `newPathStep` will replace your `pathStepProperties`. 

This `newPathStep` should be set at 2 moments:
- if you're handling a track, then you can set the pathStep at the end of `handleTrack` (instead of setting the offset, directly create the pathStep)
- if you're handling an OP, then you should add a `useEffect` which will be fired each time the `clickedOp` or the `selectedTrack` changes. In this `useEffect`, you should only update the `newPathStep`

Can you try that ?

Some remarks:
- you should return null if `newPathStep` is not defined, maybe l.147
- the state var trackOffset should not be useful anymore, I think you can remove it

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -104,6 +104,9 @@ const Map = ({
 
   const [mapIsLoaded, setMapIsLoaded] = useState(false);
 
+  const [hoveredOperationalPointId, setHoveredOperationalPointId] = useState<string | undefined>(
+    undefined
+  );
```
```suggestion
  const [hoveredOperationalPointId, setHoveredOperationalPointId] = useState<string>();
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -64,56 +84,135 @@ function AddPathStepPopup({
       <div className="details">
         <div className="details-track">
-          {featureInfoClick.feature.properties.extensions_sncf_track_name}
+          {featureInfoClick.isOperationalPoint &&
```
To improve a bit your code, I think you can extract a small component OperationalPointPopupDetails and then call it like this: 
```
      {featureInfoClick.isOperationalPoint ?
        <OperationalPointPopupDetails
          operationalPoint={featureInfoClick}
          clickedOp={clickedOp!}
          selectedTrack={selectedTrack!}
          setSelectedTrack={setSelectedTrack}
        />
        : (
          <div className="details">
            <div className="details-track">
              <small>{featureInfoClick.feature.properties.extensions_sncf_line_code}</small>
            </div>
            <div className="details-line">
              {featureInfoClick.feature.properties.extensions_sncf_line_name}
            </div>
          </div>
        )}
```

theocrsb replied:
Done

clarani commented on front/src/reducers/osrdconf/helpers.ts:
```diff
@@ -13,26 +14,44 @@ import type { PathStep } from './types';
+
   // If one of these is missing, via is not valid (it hasn't been added via click on map) and we return the same array
-  if (!('track' in newVia) || !newVia.coordinates) return pathSteps;
+  if (!('track' in newVia) && !newVia.coordinates && !isOp) return pathSteps;
```
I think the correct condition is this one: 
The pathStep is invalid if:
  - it is not an operational point AND it is not a track
  - OR it does not have coordinates

```suggestion
  if ((!isOp && !('track' in newVia)) || !newVia.coordinates) return pathSteps;
```

theocrsb replied:
Done

clarani commented on front/src/reducers/osrdconf/helpers.ts:
```diff
@@ -13,26 +14,44 @@ import type { PathStep } from './types';
-    coordinates: newVia.coordinates,
-  };
+
+  let newStep;
+
+  if ('track' in newVia) {
+    newStep = {
+      track: newVia.track,
+      offset: newVia.offset,
+      id: newVia.id,
+      coordinates: newVia.coordinates,
+    };
+    // If the via is an OP we don't need to calculate the coordinates
+  } else {
+    newStep = {
+      uic: newVia.uic,
+      secondary_code: newVia.secondary_code,
+      id: newVia.id,
+      coordinates: newVia.coordinates,
+      track_reference: newVia.track_reference,
+    };
+  }
```
nit :wink: 
```suggestion
 const newStep = {
    id: newVia.id,
    coordinates: newVia.coordinates,
    ...('track' in newVia
      ? {
          track: newVia.track,
          offset: newVia.offset,
        }
      : {
          uic: newVia.uic,
          secondary_code: newVia.secondary_code,
          track_reference: newVia.track_reference,
        }),
  };
```

theocrsb replied:
Done

clarani commented on front/src/modules/pathfinding/helpers/getStepLocation.ts:
```diff
@@ -1,19 +1,35 @@
   }
   if ('operational_point' in step) {
-    return { operational_point: step.operational_point };
+    return { operational_point: step.operational_point, track_reference: trackReference };
```
Once you are here, you have already handle the case `'track' in step`, so you can directly use `step.track_reference`
```suggestion
    return { operational_point: step.operational_point, track_reference: step.track_reference };
```

You can do the same below and remove the const `trackReference`

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/OperationalPointPopupDetails.tsx:
```diff
@@ -0,0 +1,60 @@
+  >;
+};
+
+const OperationalPointPopupDetails: React.FC<OperationalPointPopupDetailsProps> = ({
```
Don't use React.FC please :pray: 
```suggestion
const OperationalPointPopupDetails = ({
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -141,16 +142,19 @@ const Map = ({
       result.feature.properties &&
       result.feature.properties.id &&
-      result.feature.geometry.type === 'LineString'
+      (result.feature.geometry.type === 'LineString' || result.feature.geometry.type === 'Point') &&
+      (result.feature.sourceLayer === 'operational_points' ||
+        result.feature.sourceLayer === 'track_sections')
```
Since you're using again the interactive layers' id, you can remove this 3 conds
```suggestion
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -160,13 +164,17 @@ const Map = ({
   const onMoveGetFeature = (e: MapLayerMouseEvent) => {
     if (preventPointSelection) return;
-    const result = getMapMouseEventNearestFeature(e, { layersId: ['chartis/tracks-geo/main'] });
+    const result = getMapMouseEventNearestFeature(e);
```
Same here

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/ManageTrainScheduleMap/AddPathStepPopup.tsx:
```diff
@@ -64,56 +84,123 @@ function AddPathStepPopup({
+    }
+  }, [clickedOp, selectedTrack]);
+
+  if (!newPathStep || !featureInfoClick.feature.properties) return null;
```
I have tested a bit and if I click on an Op, and then select a new Op, I have an error. I think you should add a small check here ` || (featureInfoClick.isOperationalPoint && !clickedOp)`.

Also in the useEffect above, if `clickedOp` or `selectedTrack` is not defined, maybe we should reset the pathStep to undefined ?

theocrsb replied:
Done

clarani commented on front/src/modules/pathfinding/helpers/getStepLocation.ts:
```diff
@@ -1,19 +1,35 @@
+    return {
+      trigram: step.trigram,
+      secondary_code: step.secondary_code,
+      track_reference: trackReference,
```
```suggestion
      track_reference: step.track_reference,
```

theocrsb replied:
Done

clarani commented on front/src/modules/pathfinding/helpers/getStepLocation.ts:
```diff
@@ -1,19 +1,35 @@
+  return {
+    uic: step.uic,
+    secondary_code: step.secondary_code,
+    track_reference: trackReference,
```
```suggestion
    track_reference: step.track_reference,
```

theocrsb replied:
Done

theocrsb committed:
front: fix duplicate op in timestops inputs

comparing ops and pathsteps with their uic and ch code is sometimes not enough.
compare them also with position on path to differenciate them.

Signed-off-by: theocrsb <theo_crosbie@yahoo.fr>

clarani commented on front/src/modules/pathfinding/helpers/getStepLocation.ts:
```diff
@@ -1,19 +1,35 @@
 import { mToMm } from 'utils/physics';
 
 const getStepLocation = (step: PathItemLocation): PathItemLocation => {
+  const trackReference: OperationalPointReference['track_reference'] =
+    !('track' in step) && step.track_reference && 'track_name' in step.track_reference
+      ? { track_name: step.track_reference.track_name }
+      : null;
```
You can remove this now 
```suggestion
```

theocrsb replied:
Done

clarani reviewed (approved):
LGTM :white_check_mark:

clarani reviewed (approved):
LGTM, but I can't test :white_check_mark:

clarani reviewed (approved):
LGTM :white_check_mark:

Uriel-Sautron reviewed (approved):
Good work, LGTM tested !!!

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -146,16 +148,18 @@ const Map = ({
+    const result = getMapMouseEventNearestFeature(e, {
+      layersId: [
+        'chartis/tracks-geo/main',
+        ...(layersSettings.operationalpoints ? OPERATIONAL_POINT_LAYERS : ''),
```
```suggestion
        ...(layersSettings.operationalpoints ? OPERATIONAL_POINT_LAYERS : []),
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Map.tsx:
```diff
@@ -165,13 +169,22 @@ const Map = ({
+    const result = getMapMouseEventNearestFeature(e, {
+      layersId: [
+        'chartis/tracks-geo/main',
+        ...(layersSettings.operationalpoints ? OPERATIONAL_POINT_LAYERS : ''),
```
```suggestion
        ...(layersSettings.operationalpoints ? OPERATIONAL_POINT_LAYERS : []),
```

theocrsb replied:
Done

clarani commented on front/src/modules/trainschedule/components/ManageTrainSchedule/Const.ts:
The file is usually named "consts.ts" ;)
- no maj
- add a "s" at the end

theocrsb replied:
Done
